test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/RequestTests.cs

```diff
@@ -184,6 +188,96 @@ public Task RemoteIPv6Address()
             }
         }
 
+        [Fact]
+        public async Task ConnectionResetAbortsRequest()
+        {
+            await Task.Run(() =>
+            {
+                const int connectionErrorEventId = 14; // from KestrelTrace.cs
+
+                var testSink = new TestSink(write => write.EventId.Id == connectionErrorEventId);
+                var builder = new WebHostBuilder()
+                    .UseLoggerFactory(new TestLoggerFactory(testSink, true))
+                    .UseKestrel()
+                    .UseUrls($"http://127.0.0.1:0")
+                    .Configure(app => app.Run(context =>
+                    {
+                        return Task.FromResult(0);
+                    }));
+
+                using (var host = builder.Build())
+                {
+                    host.Start();
+
+                    using (var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+                    {
+                        socket.Connect(new IPEndPoint(IPAddress.Loopback, host.GetPort()));
+                        socket.LingerState = new LingerOption(true, 0);
+                    }
+
+                    // Give it some time for the connection error log message to be logged
+                    var filteredWrites = testSink.Writes.Where(write => write.EventId.Id == connectionErrorEventId);
+                    while (!filteredWrites.Any()) ;
+
+                    var connectionErrorMessage = filteredWrites.FirstOrDefault();
+                    Assert.NotNull(connectionErrorMessage);
+                    Assert.Contains("ECONNRESET", connectionErrorMessage.Exception.Message);
+                }
+            }).TimeoutAfter(TimeSpan.FromMilliseconds(2500));
+        }
+
+        [Fact]
+        public async Task ThrowsOnReadAfterConnectionError_RequestBody()
+        {
+            var requestStarted = new SemaphoreSlim(0);
+            var connectionReset = new SemaphoreSlim(0);
+            var appDone = new SemaphoreSlim(0);
+            var ioExceptionThrown = false;
+
+            var builder = new WebHostBuilder()
+                .UseKestrel()
+                .UseUrls($"http://127.0.0.1:0")
+                .Configure(app => app.Run(async context =>
+                {
+                    requestStarted.Release();
+                    await connectionReset.WaitAsync();
+
+                    try
+                    {
+                        await context.Request.Body.ReadAsync(new byte[1], 0, 1);
+                    }
+                    catch (BadHttpRequestException)
+                    {
+                        // We need this here because BadHttpRequestException derives from IOException,
+                        // and we're looking for an actual IOException.
+                    }
+                    catch (IOException)
+                    {
+                        ioExceptionThrown = true;
```

Comment:
You could capture the exception and check that the inner exception is a `TaskCanceledException` with a "The request was aborted" message.